### PR TITLE
Dynamically register/unregister databases

### PR DESCRIPTION
**Pull request by r0mant**
This PR implements the final part of [RFD31](https://github.com/gravitational/teleport/pull/6787) and updates the database server agent with a watcher that monitors changes to `Database` resources in the cluster and registers/unregisters matching resources appropriately in its proxying configuration.

The selector is a new field in the agent config:

```yaml
db_service:
  enabled: "yes"
  selectors:
  - match_labels:
      '*': '*'
```

Dynamic registration can be combined with databases defined statically in yaml configuration as well.

I've also added new `db.create`, `db.update` and `db.delete` audit events emitted for database resources. I'll update UI to recognize them later.

Refs https://github.com/gravitational/teleport/issues/6538.

### CODE REVIEW

**Comment by r0mant** [comment]
Converted to draft for the time being, need to make one additional change.

File: lib/reversetunnel/srv.go [hunk 1]
```diff
@@ -1035,19 +1035,17 @@ func newRemoteSite(srv *server, domainName string, sconn ssh.Conn) (*remoteSite,
 	}
 	remoteSite.remoteClient = clt
 
-	// DELETE IN: 8.0.0
-	//
 	// Check if the cluster that is connecting is a pre-v7 cluster. If it is,
```
Comment (zmb3) [inline lib/reversetunnel/srv.go:1038]:
```suggestion
	// Check if the cluster that is connecting is a pre-v8 cluster. If it is,
```

**Review by timothyb89 — APPROVED** [review]

File: lib/srv/db/watcher.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			}
+		} else if db.Origin() == types.OriginConfigFile || db.Origin() == "" {
+			// Do not overwrite databases from static configuration.
+			s.log.Infof("%v is part of static configuration, not registering %v.", db, new)
+			continue
+		} else if new.GetResourceID() != db.GetResourceID() {
+			// If labels were updated, the database may no longer match.
+			if services.MatchDatabase(s.cfg.Selectors, new) {
+				s.log.Infof("%v updated, re-registering.", new)
+				if err := s.registerDatabase(ctx, new); err != nil {
```
Comment (smallinsky) [inline lib/srv/db/watcher.go:105]: Does calling registerDatabase on already registered database work and resources are cleaned  properly ?

File: lib/srv/db/watcher.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+					return trace.Wrap(err)
+				}
+			} else {
+				s.log.Infof("%v updated and no longer matches, unregistering.", new)
+				if err := s.unregisterDatabase(ctx, new.GetName()); err != nil {
+					return trace.Wrap(err)
+				}
+			}
+		} else {
+			s.log.Debugf("%v is already registered.", new)
```
Comment (smallinsky) [inline lib/srv/db/watcher.go:115]: Do we want to update/restart the database if `new.GetResourceID() == db.GetResourceID()`  ? For example if user updated the `database.uri` for already registered database the current flow will ignore this change.

File: api/types/types.proto [hunk 4]
```diff
@@ -207,8 +207,8 @@ message DatabaseServerSpecV3 {
     // DEPRECATED: Moved to DatabaseSpecV3. DELETE IN 9.0.
     GCPCloudSQL GCP = 11
         [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "gcp,omitempty", deprecated = true ];
-    // Databases is a list of databases this database server proxies.
-    repeated DatabaseV3 Databases = 12 [ (gogoproto.jsontag) = "databases,omitempty" ];
+    // Database is the database proxied by this database server.
+    DatabaseV3 Database = 12 [ (gogoproto.jsontag) = "database,omitempty" ];
```
[the remark this answers is not in the document]
Comment (r0mant) [reply api/types/types.proto:211]: It's not as far as I know. There hasn't been any releases/pre-releases yet with the `Databases` field. I introduced it in https://github.com/gravitational/teleport/pull/7771 which only got merged to master.

File: lib/srv/db/watcher.go (hunk 2 above)
Comment (r0mant) [reply lib/srv/db/watcher.go]: Good call, re-registering worked but it wasn't cleaned up properly. Updated to stop heartbeat and label updates when re-registering.

File: lib/srv/db/watcher.go (hunk 3 above)
Comment (r0mant) [reply lib/srv/db/watcher.go:116]: Resource ID changes when you update the resource so if you update URI, description, labels or any other resource field, the change will be detected. I've added a test case for this as well.

**Review by smallinsky — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**